File: src/etch/bindings/csharp/util/ListSerializer.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;
using Etch.Msg;
using Etch.Support;

namespace Etch.Util
{
    public class ListSerializer : ImportExportHelper
    {
        private static String FIELD_NAME = "values";

         /// <summary>
        /// Defines custom fields in the value factory so that the importer
        /// can find them.
        /// </summary>
        /// <param name="type"></param>
        /// <param name="class2type"></param>
        public static void Init( XType type, Class2TypeMap class2type )
        {
            Field field = type.GetField( FIELD_NAME );
     /*       class2type.Add( typeof( List<object> ), type );
            type.SetClass(typeof(List<object>)); */
            class2type.Add(typeof(ArrayList), type);
            type.SetComponentType(typeof(ArrayList));
            type.SetImportExportHelper( new ListSerializer( type, field ) );
            type.PutValidator( field, Validator_object.Get( 1 ) );
            type.Lock();
        }

        public ListSerializer( XType type, Field field )
        {
            this.type = type;
            this.field = field;
        }

        private readonly XType type;
        private readonly Field field;


        public override Object ImportValue( StructValue sv )
        {
       //    List<object> list = new List<object>();
            ArrayList list = new ArrayList();
            Object[] values = ( Object[] ) sv.Get( field );
            int n = values.Length;
            int index = 0;
            while ( index < n )
            {
               
                object value = ( object ) values[ index++ ];
                list.Add( value );
            }

            return list;
        }

        public override StructValue ExportValue( Object value )
        {
     //       List<object> list = (List<object>)value;
            ArrayList list = (ArrayList)value;
            Object[] values = new Object[ list.Count];
            int index = 0;

            foreach ( object me in list )
            {
                values[ index++ ] = me;
              
            }

            StructValue sv = new StructValue( type );
            sv.Add( field, values );
            return sv;

        }

    }
}

File: src/etch/bindings/java/util/ListSerializer.java
/* $Id$
 *
 * Created by sccomer on Apr 14, 2008.
 *
 * Copyright (c) 2005 Metreos, Inc. All rights reserved.
 */

package etch.bindings.java.util;

import java.util.ArrayList;
import java.util.List;

import etch.bindings.java.msg.Field;
import etch.bindings.java.msg.ImportExportHelper;
import etch.bindings.java.msg.StructValue;
import etch.bindings.java.msg.Type;
import etch.bindings.java.support.Class2TypeMap;
import etch.bindings.java.support.Validator_object;

/**
 * Serializer for generic List<Object>.
 */
public class ListSerializer implements ImportExportHelper
{
	private final static String FIELD_NAME = "values";

	/**
	 * Defines custom fields in the value factory so that the importer
	 * can find them.
	 * @param type
	 * @param class2type
	 */
	public static void init( Type type, Class2TypeMap class2type )
    {
		Field field = type.getField( FIELD_NAME );
		class2type.put( ArrayList.class, type );
		type.setComponentType( List.class );
		type.setImportExportHelper( new ListSerializer( type, field ) );
		type.putValidator( field, Validator_object.get( 1 ) );
		type.lock();
    }

	/**
	 * @param type
	 * @param field
	 */
	public ListSerializer( Type type, Field field )
	{
		this.type = type;
		this.field = field;
	}
	
	private final Type type;
	
	private final Field field;
	
	public Object importValue( StructValue struct )
	{
		struct.checkType( type );
		Object[] values = (Object[]) struct.get( field );
		
		List<Object> list = new ArrayList<Object>( values.length );
		for( Object value: values)
			list.add( value );
		
		return list;
	}

	@SuppressWarnings("unchecked")
	public StructValue exportValue( Object value )
	{
		List<Object> list = (List<Object>) value;
		
		Object[] values = new Object[list.size()];
		list.toArray( values );
		
		StructValue struct = new StructValue( type );
		struct.put( field, values );
		return struct;
	}
}
